bls12_381/src/fp.rs

```diff
@@ -545,6 +608,181 @@ impl Fp {
     }
 }
 
+pub struct FpRepr([u8; 48]);
+
+impl fmt::Debug for FpRepr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "0x")?;
+        for &b in self.0.iter() {
+            write!(f, "{:02x}", b)?;
+        }
+        Ok(())
+    }
+}
+
+impl Default for FpRepr {
+    fn default() -> Self {
+        FpRepr([0; 48])
+    }
+}
+
+impl AsRef<[u8]> for FpRepr {
+    fn as_ref(&self) -> &[u8] {
+        &self.0[..]
+    }
+}
+
+impl AsMut<[u8]> for FpRepr {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0[..]
+    }
+}
+
+impl From<Fp> for FpRepr {
+    fn from(value: Fp) -> FpRepr {
+        value.into_repr()
+    }
+}
+
+impl<'a> From<&'a Fp> for FpRepr {
+    fn from(value: &'a Fp) -> FpRepr {
+        value.into_repr()
+    }
+}
+
+impl From<u64> for Fp {
+    fn from(val: u64) -> Fp {
+        Fp([val, 0, 0, 0, 0, 0]) * R2
+    }
+}
+
+impl BitAnd<u64> for Fp {
+    type Output = u64;
+
+    fn bitand(self, rhs: u64) -> u64 {
+        let tmp = Fp::montgomery_reduce(
+            self.0[0], self.0[1], self.0[2], self.0[3], self.0[4], self.0[5], 0, 0, 0, 0, 0, 0,
+        );
+        tmp.0[0] & rhs
+    }
+}
+
+impl Shr<u32> for Fp {
+    type Output = Self;
+
+    fn shr(self, mut n: u32) -> Self {
+        if n as usize >= 64 * 4 {
+            return Self::from(0);
+        }
+
+        // Convert from Montgomery to native representation.
+        let mut tmp = Fp::montgomery_reduce(
+            self.0[0], self.0[1], self.0[2], self.0[3], self.0[4], self.0[5], 0, 0, 0, 0, 0, 0,
+        );
+
+        while n >= 64 {
+            let mut t = 0;
+            for i in tmp.0.iter_mut().rev() {
+                mem::swap(&mut t, i);
+            }
+            n -= 64;
+        }
+
+        if n > 0 {
+            let mut t = 0;
+            for i in tmp.0.iter_mut().rev() {
+                let t2 = *i << (64 - n);
+                *i >>= n;
+                *i |= t;
+                t = t2;
+            }
+        }
+
+        // Convert back to Montgomery representation
+        tmp * R2
+    }
+}
+
+impl Field for Fp {
+    fn random<R: RngCore + ?Sized>(_rng: &mut R) -> Self {
+        unimplemented!()
+    }
+
+    fn zero() -> Self {
+        Self::zero()
+    }
+
+    fn one() -> Self {
+        Self::one()
+    }
+
+    fn is_zero(&self) -> bool {
+        self.ct_eq(&Self::zero()).into()
+    }
+
+    #[must_use]
+    fn square(&self) -> Self {
+        self.square()
+    }
+
+    #[must_use]
+    fn double(&self) -> Self {
+        self.add(self)
+    }
+
+    fn invert(&self) -> CtOption<Self> {
+        self.invert()
+    }
+
+    fn frobenius_map(&mut self, _power: usize) {
```

Comment:
I don't think we need to offer this as part of the `Field` trait since it's only ever used internally in `bls12_381`, and field extensions aren't exposed anywhere in our stack.

Reply:
Cool, we can drop this.